src/lib.rs

```diff
@@ -40,192 +40,23 @@
 
 extern crate libc;
 
-mod errors;
+pub mod errors;
+pub mod backends;
 
-use libc::{pid_t, c_char, c_void, getpid, size_t, c_int, geteuid};
-use std::ffi::CString;
-use std::path::PathBuf;
 use errors::TraceMeError;
-use std::fs::File;
-use std::io::Read;
 
-#[cfg(target_os="linux")]
-const PERF_PERMS_PATH: &'static str = "/proc/sys/kernel/perf_event_paranoid";
-
-// FFI stubs
-#[link(name = "traceme")]
-extern "C" {
-    fn traceme_start_tracer(conf: *const TracerConf) -> *const c_void;
-    fn traceme_stop_tracer(tr_ctx: *const c_void) -> c_int;
-}
-
-// Struct used to communicate a tracing configuration to C. Must stay in sync with the C code.
-#[repr(C)]
-struct TracerConf {
-    target_pid: pid_t,
-    trace_filename: *const c_char,
-    map_filename: *const c_char,
-    data_bufsize: size_t,
-    aux_bufsize: size_t,
-}
-
-/// Represents an instance of the Intel PT tracer.
-#[derive(Default)]
-pub struct Tracer {
-    /// Filename to store the trace to.
-    trace_filename: String,
-    /// PID to trace.
-    target_pid: pid_t,
-    /// Data buffer size, in pages. Must be a power of 2.
-    data_bufsize: size_t,
-    /// Aux buffer size, in pages. Must be a power of 2.
-    aux_bufsize: size_t,
-    /// Opaque C pointer representing the tracer context.
-    tracer_ctx: Option<*const c_void>,
-}
-
-impl Tracer {
-    /// Create a new tracer.
-    ///
-    /// # Example
-    ///
-    /// ```
-    /// use traceme::Tracer;
+/// The interface offered by all tracer types.
+///
+/// It is the job of the consumer of this library to decide which specific tracer type to
+/// instantiate, and to set it up using its type-specific configuration methods. Once configured a
+/// tracer can be treated as a generic data structure with Box<Tracer>.
```

Comment:
I don't think we need to mention the `Box` thing here (plus the phrasing doesn't currently make sense). It's up to users whether they want dynamic dispatch on a `Tracer` or not.

Reply:
OK, I guess you are right. Let's assume the user knows that they can't store a trait object without an additional wrapper like a `Box`.